docs: Added zero downtime tests documentation (#7895)
keptn/keptn: docs/developer/zero_downtime_tests.md

# Zero Downtime Tests

<details>
<summary>Table of Contents</summary>

<!-- toc -->

- [Zero Downtime Tests](#zero-downtime-tests)
  - [Structure of Zero Downtime Tests](#structure-of-zero-downtime-tests)
    - [Adding a new Zero Downtime Test](#adding-a-new-zero-downtime-test)
  - [Running Zero Downtime Tests](#running-zero-downtime-tests)
  - [Run Zero Downtime Tests remotely on Github](#run-zero-downtime-tests-remotely-on-github)
  - [Run Zero Downtime Tests locally](#run-zero-downtime-tests-locally)
    - [Prepare your local environment to run Zero Downtime tests](#prepare-your-local-environment-to-run-zero-downtime-tests)
    - [Run the full installation of Zero Downtime Tests locally](#run-the-full-installation-of-zero-downtime-tests-locally)

<!-- tocstop -->
</details>

## Structure of Zero Downtime Tests

The Zero Downtime Tests and their resources are located under the `/test` directory in this repository. There are two main directories:
* `/test/zero-downtime/assets` -> directory containing resources and scripts, which are used during the run of the Zero Downtime Tests 
* `/test/zero-downtime` -> Zero Downtime Tests and testsuites

> Zero Downtime Tests are organized into testsuites, a main suite to run all tests, andd a separate file for each specific scenario:

Reviewer: ```suggestion
Zero Downtime Tests are organized into testsuites, a main suite to run all tests, and a separate file for each specific scenario:
```

> * testsuite_zerodowntime_test contains the Test_ZeroDowntime function that runs all test suites, each test is based on the ZDTestTemplate that runs a test function in parallel to rolling upgrade and api probing

Reviewer: ```suggestion
* testsuite_zerodowntime_test contains the Test_ZeroDowntime function that runs all test suites, each test is based on the ZDTestTemplate that runs a test function in parallel to rolling upgrade and API probing
```

* sequence_test contains test for default sequences: evaluation and deployment 
* API_test contains API probes of "indempotent" endpoints to be run nonstop to verify keptn API availability during upgrades 
* webhook_test contains tests for sequences calling webhooks

> The api tests are run in parallel during pipeline execution on Github, webhook and sequences are passed to the  ZDTemplate and are run sequentially.

Reviewer: ```suggestion
API tests are run in parallel during pipeline execution on Github, webhook and sequences are passed to the  ZDTemplate and are run sequentially.
```

> ### Adding a new Zero DowntimeTest

Reviewer: ```suggestion
### Adding a new Zero Downtime Test
```

Adding a new Zero DowntimeTest consists of two steps:

> 1. Copy the webhook test suite file in `/test/zero-downtime` directory as a reference. Each Test_ function will be run sequentially, SetupSuite will run once at suite creation, to decide the behaviour of the test suite during upgrades edit

Reviewer: ```suggestion
1. Copy the webhook test suite file in `/test/zero-downtime` directory as a reference. Each `Test_` function will be run sequentially, `SetupSuite` will run once at suite creation. To change the behavior of the test suite during upgrades edit
```

    ```
    func Webhook(t *testing.T, env *ZeroDowntimeEnv) 
    ```

>    This is currently set to run continuously a new suite until the upgrades are terminated

Reviewer: ```suggestion
   This is currently set to run a new suite continuously until the upgrades are terminated
```

> 4. Add the new created function to the testsuite_zerodowntime_test like so 

Reviewer: ```suggestion
2. Add the new created function to the `testsuite_zerodowntime_test` like so 
```

These two options can be also combined and currently, only executions of all Zero Downtime Tests for all testsuites are supported. The execution of the Tests is fairly easy:
1. Navigate to the `Actions` tab in the `keptn/keptn` repository (https://github.com/keptn/keptn)
2. Choose `Zero Downtime Tests` from the left side menu
3. Click on `Run Workflow`, where a dialog window will appear. 
   Here, you need to choose the context (`Use Workflow from`) of the tests you wish to use (`master` is the default). 
   You should use this `master` context unless you have not made any changes in the Zero Downtime Tests pipeline. 
   Secondly, you choose a branch, from which the CI build artifacts (docker images) should be used.
   Here, you mostly use the branch of the code you are currently working on and want to run Zero Downtime Tests for your code changes. Please be aware, that you need to wait for the docker images to be built before you can execute the Integartion Tests.
   You have then to pass the image tag for the upgrade command e.g. 0.15.1-dev or a PR one like 0.15.0-PR-6897.

>    If you want to perform ui test thick the first box, if you prefer to delete the generated cluster immediately tick the second box.

Reviewer: ```suggestion
   If you want to perform UI test tick the first box. If you prefer to delete the generated cluster immediately tick the second box.
```
